The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5030136758/jobs/9022354255) [(plain)](https://github.com/rust-lang/rust/commit/5069d40aba51f3931e86519e4ebd33a98c400feb/checks/13624924537/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling clap_complete v4.2.2
error[E0308]: mismatched types
    --> lib.rs:1327:17
     |
1326 |             "beta" => match self.beta_prerelease_version() {
     |                             ------------------------------ this expression has type `u32`
1327 |                 Some(revision) if !self.config.omit_git_hash => {
     |                 ^^^^^^^^^^^^^^ expected `u32`, found `Option<_>`
     = note: expected type `u32`
                found enum `std::option::Option<_>`

For more information about this error, try `rustc --explain E0308`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0308]: mismatched types
    --> lib.rs:1327:17
     |
1326 |             "beta" => match self.beta_prerelease_version() {
     |                             ------------------------------ this expression has type `u32`
1327 |                 Some(revision) if !self.config.omit_git_hash => {
     |                 ^^^^^^^^^^^^^^ expected `u32`, found `Option<_>`
     = note: expected type `u32`
                found enum `std::option::Option<_>`

For more information about this error, try `rustc --explain E0308`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0308]: mismatched types
    --> lib.rs:1327:17
     |
1326 |             "beta" => match self.beta_prerelease_version() {
     |                             ------------------------------ this expression has type `u32`
1327 |                 Some(revision) if !self.config.omit_git_hash => {
     |                 ^^^^^^^^^^^^^^ expected `u32`, found `Option<_>`
     = note: expected type `u32`
                found enum `std::option::Option<_>`

For more information about this error, try `rustc --explain E0308`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0308]: mismatched types
    --> lib.rs:1327:17
     |
1326 |             "beta" => match self.beta_prerelease_version() {
     |                             ------------------------------ this expression has type `u32`
1327 |                 Some(revision) if !self.config.omit_git_hash => {
     |                 ^^^^^^^^^^^^^^ expected `u32`, found `Option<_>`
     = note: expected type `u32`
                found enum `std::option::Option<_>`

For more information about this error, try `rustc --explain E0308`.
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0308]: mismatched types
    --> lib.rs:1327:17
     |
1326 |             "beta" => match self.beta_prerelease_version() {
     |                             ------------------------------ this expression has type `u32`
1327 |                 Some(revision) if !self.config.omit_git_hash => {
     |                 ^^^^^^^^^^^^^^ expected `u32`, found `Option<_>`
     = note: expected type `u32`
                found enum `std::option::Option<_>`

For more information about this error, try `rustc --explain E0308`.
```

</details>

--- a/src/bootstrap/builder/tests.rs
+++ b/src/bootstrap/builder/tests.rs
@@ -146,6 +146,22 @@ fn alias_and_path_for_library() {
     );
 }
 
+#[test]
+fn test_beta_rev_parsing() {
+    use crate::extract_beta_rev;
+
+    // single digit revision
+    assert_eq!(extract_beta_rev("1.99.9-beta.7 (xxxxxx)"), Some("7".to_string()));
+    // multiple digits
+    assert_eq!(extract_beta_rev("1.99.9-beta.777 (xxxxxx)"), Some("777".to_string()));
+    // nightly channel (no beta revision)
+    assert_eq!(extract_beta_rev("1.99.9-nightly (xxxxxx)"), None);
+    // stable channel (no beta revision)
+    assert_eq!(extract_beta_rev("1.99.9 (xxxxxxx)"), None);
+    // invalid string
+    assert_eq!(extract_beta_rev("invalid"), None);
+}
+
 mod defaults {
     use super::{configure, first, run_build};
     use crate::builder::*;

--- a/src/bootstrap/lib.rs
+++ b/src/bootstrap/lib.rs
@@ -1324,7 +1324,7 @@ impl Build {
         match &self.config.channel[..] {
             "stable" => num.to_string(),
             "beta" => {
-                if self.rust_info().is_managed_git_subrepository() && !self.config.omit_git_hash {
+                if !self.config.omit_git_hash {
                     format!("{}-beta.{}", num, self.beta_prerelease_version())
                 } else {
                     format!("{}-beta", num)
@@ -1336,18 +1336,28 @@ impl Build {
     }
 
     fn beta_prerelease_version(&self) -> u32 {
+        fn extract_beta_rev_from_file<P: AsRef<Path>>(version_file: P) -> Option<String> {
+            let version = fs::read_to_string(version_file).ok()?;
+
+            extract_beta_rev(&version)
+        }
+
         if let Some(s) = self.prerelease_version.get() {
             return s;
         }
 
-        // Figure out how many merge commits happened since we branched off master.
-        // That's our beta number!
-        // (Note that we use a `..` range, not the `...` symmetric difference.)
-        let count =
+        // First check if there is a version file available.
+        // If available, we read the beta revision from that file.
+        // This only happens when building from a source tarball when Git should not be used.
+        let count = extract_beta_rev_from_file(self.src.join("version")).unwrap_or_else(|| {
+            // Figure out how many merge commits happened since we branched off master.
+            // That's our beta number!
+            // (Note that we use a `..` range, not the `...` symmetric difference.)
             output(self.config.git().arg("rev-list").arg("--count").arg("--merges").arg(format!(
                 "refs/remotes/origin/{}..HEAD",
                 self.config.stage0_metadata.config.nightly_branch
-            )));
+            )))
+        });
         let n = count.trim().parse().unwrap();
         self.prerelease_version.set(Some(n));
         n
@@ -1705,6 +1715,17 @@ to download LLVM rather than building it.
         stream.reset().unwrap();
         result
     }
+}
+
+/// Extract the beta revision from the full version string.
+///
+/// The full version string looks like "a.b.c-beta.y". And we need to extract
+/// the "y" part from the string.
+pub fn extract_beta_rev(version: &str) -> Option<String> {
+    let parts = version.splitn(2, "-beta.").collect::<Vec<_>>();
+    let count = parts.get(1).and_then(|s| s.find(' ').map(|p| (&s[..p]).to_string()));
+
+    count
 }
 
 #[cfg(unix)]

